### PR TITLE
fix: handleOrderBook use safeNumber instead of safeFloat

### PR DESCRIPTION
**Pull request by samgermain**
I tested all the other exchanges, in ts and python, but just didn't copy the results

fixes: #17542 

```
2023-04-17T23:35:56.206Z
Node.js: v18.15.0
CCXT v3.0.64
ascendex.watchOrderBook (BTC/USDT)
OrderBook {
  bids: Bids(138) [
    [ 29435.44, 0.0047 ],  [ 29423.06, 0.331 ],   [ 29423.05, 0.077 ],
    ...
    [ 10000, 0.05 ],       [ 9000, 0.05 ],        [ 5888, 6.24897 ]
  ],
  asks: Asks(148) [
    [ 29450.79, 0.0068 ],  [ 29450.8, 0.077 ],    [ 29459.78, 0.01323 ],
    ...
    [ 49250, 0.0225 ],     [ 49547, 0.00038 ],    [ 73492, 0.00009 ],
    [ 100000, 0.201 ]
  ],
  timestamp: 1681774567845,
  datetime: '2023-04-17T23:36:07.845Z',
  nonce: 216348491,
  symbol: 'BTC/USDT'
}
OrderBook {
  bids: Bids(138) [
    [ 29435.44, 0.0047 ],  [ 29423.06, 0.331 ],   [ 29423.05, 0.077 ],
    ...
    [ 10000, 0.05 ],       [ 9000, 0.05 ],        [ 5888, 6.24897 ]
  ],
  asks: Asks(149) [
    [ 29450.79, 0.0068 ],  [ 29450.8, 0.077 ],    [ 29459.78, 0.01323 ],
    ...
    [ 73492, 0.00009 ],    [ 100000, 0.201 ]
  ],
  timestamp: 1681774567896,
  datetime: '2023-04-17T23:36:07.896Z',
  nonce: 216348492,
  symbol: 'BTC/USDT'
}
```

```
2023-04-18T02:15:31.946Z
Node.js: v18.15.0
CCXT v3.0.64
binance.watchOrderBook (BTC/USDT)
OrderBook {
  bids: Bids(1000) [
    [ 29405.8, 3.89945 ],   [ 29405.75, 0.00689 ], [ 29405.74, 0.00669 ],
    ...
    [ 29258.06, 0.00035 ],  [ 29258, 0.00683 ],    [ 29257.84, 0.00058 ],
    [ 29257.62, 0.00046 ]
  ],
  asks: Asks(1000) [
    [ 29405.81, 13.21899 ], [ 29405.82, 0.07735 ], [ 29405.84, 1.2 ],
    ...
    [ 29490.32, 0.00051 ],  [ 29490.55, 0.00057 ], [ 29490.59, 0.0102 ],
    [ 29490.77, 0.00088 ]
  ],
  timestamp: undefined,
  datetime: undefined,
  nonce: 36269916090,
  symbol: 'BTC/USDT'
}
OrderBook {
  bids: Bids(1000) [
    [ 29405.8, 3.89945 ],  [ 29405.75, 0.00689 ], [ 29405.74, 0.00669 ],
    ...
    [ 29258.26, 0.00795 ], [ 29258.17, 0.00239 ], [ 29258.06, 0.00035 ],
    [ 29258, 0.00683 ]
  ],
  asks: Asks(1000) [
    [ 29405.81, 13.22471 ], [ 29405.82, 0.07735 ], [ 29405.84, 1.2 ],
    ...
    [ 29490.32, 0.00051 ],  [ 29490.55, 0.00057 ], [ 29490.59, 0.0102 ],
    [ 29490.77, 0.00088 ]
  ],
  timestamp: 1681784139281,
  datetime: '2023-04-18T02:15:39.281Z',
  nonce: 36269916098,
  symbol: 'BTC/USDT'
}
```

```
Python v3.11.2
CCXT v3.0.64
binance.watchOrderBook(BTC/USDT)
{'asks': [[29431.03, 4.68929], [29431.07, 0.28333], ... [29516.09, 0.00428]],
 'bids': [[29431.02, 11.67975], [29431.01, 0.00068], ... [29291.15, 0.00091]],
 'datetime': None,
 'nonce': 36270040565,
 'symbol': 'BTC/USDT',
 'timestamp': None}
{'asks': [[29431.03, 4.68929], [29431.07, 0.28333], ... [29516.09, 0.00428]],
 'bids': [[29431.02, 11.67975], [29431.01, 0.00068], ... [29291.15, 0.00091]],
 'datetime': '2023-04-18T02:28:11.681Z',
 'nonce': 36270040566,
 'symbol': 'BTC/USDT',
 'timestamp': 1681784891681}
```

```
Python v3.11.2
CCXT v3.0.64
zb.watchOrderBook(BTC/USDT)
{'asks': [[29371.84, 0.0163], [29389.42, 0.0152], ... [29442.15, 0.018]],
 'bids': [[29000.01, 0.0234], [29000.0, 0.1305], ... [28899.0, 0.0065]],
 'datetime': '2023-04-18T02:10:50.426Z',
 'nonce': None,
 'symbol': 'BTC/USDT',
 'timestamp': 1681783850426}
{'asks': [[29371.84, 0.0163], [29389.42, 0.0152], ... [29442.15, 0.018]],
 'bids': [[29000.01, 0.0234], [29000.0, 0.1305], ... [28899.0, 0.0065]],
 'datetime': '2023-04-18T02:10:50.526Z',
 'nonce': None,
 'symbol': 'BTC/USDT',
 'timestamp': 1681783850526}
```

```
2023-04-18T02:22:01.864Z
Node.js: v18.15.0
CCXT v3.0.64
mexc.watchOrderBook (BTC/USDT)
OrderBook {
  bids: Bids(100) [
    [ 29447.79, 2.499899 ], [ 29447.78, 0.58512 ],  [ 29447.77, 1.606202 ],
    ...
    [ 28637.43, 0.001644 ], [ 28630, 0.013071 ],    [ 28620, 0.001642 ],
    [ 28600, 0.002048 ]
  ],
  asks: Asks(100) [
    [ 29448.1, 2.251793 ],  [ 29448.12, 2.771646 ], [ 29448.13, 1.047322 ],
    ...
    [ 30521.5, 0.000419 ],  [ 30523, 0.159359 ],    [ 30536, 0.015362 ],
    [ 30546.08, 0.0025 ]
  ],
  timestamp: undefined,
  datetime: undefined,
  nonce: 14482090682,
  symbol: 'BTC/USDT'
}
OrderBook {
  bids: Bids(101) [
    [ 29447.8, 2.042705 ],  [ 29447.79, 2.499899 ], [ 29447.78, 0.58512 ],
    ...
    [ 28650, 0.06638 ],     [ 28637.43, 0.001644 ], [ 28630, 0.013071 ],
    [ 28620, 0.001642 ],    [ 28600, 0.002048 ]
  ],
  asks: Asks(100) [
    [ 29448.1, 2.251793 ],  [ 29448.12, 2.771646 ], [ 29448.13, 1.047322 ],
    ...
    [ 30521.5, 0.000419 ],  [ 30523, 0.159359 ],    [ 30536, 0.015362 ],
    [ 30546.08, 0.0025 ]
  ],
  timestamp: 1681784528523,
  datetime: '2023-04-18T02:22:08.523Z',
  nonce: 14482090683,
  symbol: 'BTC/USDT'
}
```

```
Python v3.11.2
CCXT v3.0.64
mexc.watchOrderBook(BTC/USDT)
{'asks': [[29428.94, 10.662029], [29428.95, 8.482876],  ... [30609.0, 0.0021]],
 'bids': [[29428.63, 2.227123], [29428.61, 2.766185], ... [28620.0, 0.001642]],
 'datetime': None,
 'nonce': 14482178763,
 'symbol': 'BTC/USDT',
 'timestamp': None}
{'asks': [[29428.94, 10.662029], [29428.95, 8.482876],  ... [30609.0, 0.0021]],
 'bids': [[29428.63, 2.227123], [29428.61, 2.766185], ... [28620.0, 0.001642]],
 'datetime': '2023-04-18T02:29:40.871Z',
 'nonce': 14482178764,
 'symbol': 'BTC/USDT',
 'timestamp': 1681784980871}
```

### CODE REVIEW

File: ts/src/pro/ascendex.ts [hunk 1]
```diff
@@ -313,8 +313,8 @@ export default class ascendex extends ascendexRest {
         //
         // ["40990.47","0.01619"],
         //
-        const price = this.safeFloat (delta, 0);
-        const amount = this.safeFloat (delta, 1);
+        const price = this.safeNumber (delta, 0);
```
Comment (carlosmiei) [inline ts/src/pro/ascendex.ts:316]: @samgermain I don't think this is enough,internally the orderbookside uses bisect left that is not expecting a string there
image

@frosty00 can you confirm?

Comment (samgermain) [reply ts/src/pro/ascendex.ts:316]: To test this out, do I just need to set 

```
"options": {
            "number": "String"
 }
```

within my `keys.local.json`?

Comment (samgermain) [reply ts/src/pro/ascendex.ts:316]: @carlosmiei It works fine actually

This script

```ts
import ascendex from '/Users/samgermain/Documents/CCXT/ccxt/ts/src/pro/ascendex.js';
import keys from './keys.local.json' assert { type: "json" };

async function main () {

    const ascendexEx = new ascendex ({'apiKey': keys.ascendex.apiKey, 'secret': keys.ascendex.secret});
    // @ts-ignore
    ascendexEx.number = String; // @ts-ignore

    await ascendexEx.loadMarkets ();
    const ascendexBook = await ascendexEx.watchOrderBook ('BTC/USDT');
    console.log (ascendexBook);

};

main ();
```

prints

```
% node --loader ts-node/esm example.ts
OrderBook {
  bids: Bids(136) [
    [ '29539.99', '0.19602' ],
    [ '29537.22', '0.0068' ],
    ...
    [ '30455', '1' ],
    [ '30470', '1' ],
    ... 203 more items
  ],
  timestamp: 1691617816607,
  datetime: '2023-08-09T21:50:16.607Z',
  nonce: 8077974,
  symbol: 'BTC/USDT'
}
```

**Comment by samgermain** [comment]
fixed in #18920

**Pull request closed without merging**